src/data/bucket/line_bucket.js

```diff
@@ -110,20 +124,78 @@ class LineBucket implements Bucket {
+        this.features = [];
 
         this.layoutVertexArray = new LineLayoutArray();
         this.indexArray = new TriangleIndexArray();
         this.programConfigurations = new ProgramConfigurationSet(layoutAttributes, options.layers, options.zoom);
         this.segments = new SegmentVector();
+
+        this.dataDrivenPatternLayers = [];
+
+
+        for (let i = 0; i < this.layerIds.length; i++) {
+            const id = this.layerIds[i];
+            const programConfiguration = this.programConfigurations.get(id);
+            if (programConfiguration.binders && programConfiguration.binders['line-pattern'] && programConfiguration.binders['line-pattern'].isDataDriven()) {
+                this.dataDrivenPatternLayers.push(i);
+            }
+        }
     }
 
     populate(features: Array<IndexedFeature>, options: PopulateParameters) {
+        const icons = options.iconDependencies;
+        this.features = [];
+
+        // add all icons needed for this tile to the tile's IconAtlas dependencies
+        for (const layer of this.layers) {
+            const linePattern = layer.paint.get('line-pattern');
+            const image = linePattern.constantOr(null);
+            if (image) {
+                icons[image.min] = true;
+                icons[image.mid] = true;
+                icons[image.max] = true;
+            }
+        }
+
         for (const {feature, index, sourceLayerIndex} of features) {
-            if (this.layers[0]._featureFilter({zoom: this.zoom}, feature)) {
-                const geometry = loadGeometry(feature);
-                this.addFeature(feature, geometry);
-                options.featureIndex.insert(feature, geometry, index, sourceLayerIndex, this.index);
+            if (!this.layers[0]._featureFilter({zoom: this.zoom}, feature)) continue;
+            for (let i = 0; i < this.dataDrivenPatternLayers.length; i++) {
+                const layerIdx = this.dataDrivenPatternLayers[i];
+                const layer = this.layers[layerIdx];
+                const linePattern = layer.paint.get('line-pattern');
+                const image = linePattern.evaluate(feature);
+                if (image) {
+                    icons[image.min] = true;
+                    icons[image.mid] = true;
+                    icons[image.max] = true;
+                }
+            }
+
+            const geometry = loadGeometry(feature);
+            const lineFeature: LineFeature = {
+                sourceLayerIndex: sourceLayerIndex,
+                index: index,
+                geometry: geometry,
+                properties: feature.properties,
+                type: feature.type
+            };
+
+            if (typeof feature.id !== 'undefined') {
+                lineFeature.id = feature.id;
             }
+
+            this.features.push(lineFeature);
+            options.featureIndex.insert(feature, geometry, index, sourceLayerIndex, this.index);
+        }
+    }
+
+    // used if line-pattern is data-driven
```

Comment:
Isn't `addFeatures()` also used if line-pattern is constant?

Reply:
yep, sorry – an artifact from an earlier iteration 🙈